### PR TITLE
[WIP] Feature/merkle mask in coda base ledger

### PR DESCRIPTION
**Pull request by bkase**
TODO fill this out (for now just force CI to run)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -5,3 +5,5 @@ result
 test.log
 build.log
 src/config.mlh
+src/external/ocaml-rocksdb/rocksdb
```
Comment (imeckler) [inline .gitignore]: are we sure we want to ignore? Did you make modifications that are needed?

File: src/lib/coda_base/account.ml [hunk 2]
```diff
@@ -127,9 +127,14 @@ let fold_bits ({public_key; balance; nonce; receipt_chain_hash; delegate} : t)
   +> Receipt.Chain_hash.fold receipt_chain_hash
   +> Public_key.Compressed.fold delegate
 
-let hash_prefix = Hash_prefix.account
+let crypto_hash_prefix = Hash_prefix.account
 
-let hash t = Pedersen.hash_fold hash_prefix (fold_bits t)
+let crypto_hash t = Pedersen.hash_fold crypto_hash_prefix (fold_bits t)
+
+let empty_crypto_hash =
```
Comment (imeckler) [inline src/lib/coda_base/account.ml]: I don't think this is used.

File: src/lib/coda_base/account.ml [hunk 3]
```diff
@@ -138,11 +143,9 @@ let empty =
   ; receipt_chain_hash= Receipt.Chain_hash.empty
   ; delegate= Public_key.Compressed.empty }
 
-let digest t = Pedersen.State.digest (hash t)
-
-let empty_hash = digest empty
+let digest t = Pedersen.State.digest (crypto_hash t)
 
-let pubkey t = t.public_key
+let public_key_of_account (t : t) = t.public_key
```
Comment (imeckler) [inline src/lib/coda_base/account.ml]: can we call this `public_key`?

File: src/lib/coda_base/ledger.ml [hunk 4]
```diff
@@ -2,416 +2,510 @@ open Core
...
+
+type unattached_mask = Mask.t
+
+type attached_mask = Mask.Attached.t
+
+(* inside MaskedLedger, the functor argument has assigned to location, account, and path
+   but the module signature for the functor result wants them, so we declare them here *)
+type location = Location.t
+
+let fold_until t ~init ~f ~finish =
```
Comment (imeckler) [inline src/lib/coda_base/ledger.ml]: can we make an issue for this to not allocate a list? I feel like the expectation with a fold is that e.g. `fold t ~f:(fun acc _ -> acc + 1) ~init:0` should use a constant amount of memory

File: src/lib/coda_base/ledger.ml [hunk 5]
```diff
@@ -2,416 +2,510 @@ open Core
...
+          (UC.check tr.common.user_command) ~f:(fun x ->
+            Ok (Transaction.User_command x) )
+    | Fee_transfer f -> Ok (Fee_transfer f.fee_transfer)
+    | Coinbase c -> Ok (Coinbase c.coinbase)
+end
+
+(* someday: It would probably be better if we didn't modify the receipt chain hash
+ in the case that the sender is equal to the receiver, but it complicates the SNARK, so
+ we don't for now. *)
+let apply_user_command_unchecked ledger
```
Comment (imeckler) [inline src/lib/coda_base/ledger.ml]: I assume you didnt change the logic of any of these functions yes? just a bad diff?

File: src/lib/ledger_builder_controller/ledger_builder_controller.ml [hunk 6]
```diff
@@ -123,45 +124,20 @@ end = struct
...
-        {data= genesis_tip; hash= genesis_state_hash}
-    | Error `Checksum_no_match ->
-        Logger.error log
-          !"Checksum from location %s does not match with data\n\
-            Using Genesis tip"
-          longest_tip_location ;
-        {data= genesis_tip; hash= genesis_state_hash}
+  let load_tip_and_genesis_hash {Config.genesis_tip; _} log =
+    let genesis_state_hash = Protocol_state.hash genesis_tip.state in
+    Logger.info log
```
Comment (imeckler) [inline src/lib/ledger_builder_controller/ledger_builder_controller.ml]: make an issue to fix this asap

File: src/lib/merkle_ledger/any_ledger.ml [hunk 7]
```diff
@@ -14,16 +14,46 @@
...
+module type S = sig
+  type key
+
+  type account
+
+  type hash
+
+  module Location : Location_intf.S
+
+  type witness [@@deriving sexp_of]
```
Comment (imeckler) [inline src/lib/merkle_ledger/any_ledger.ml]: why witness?

File: src/lib/merkle_ledger/any_ledger.ml [hunk 8]
```diff
@@ -14,16 +14,46 @@
...
+  module type Base_intf =
+    Base_ledger_intf.S
+    with module Addr = Location.Addr
+    with module Location = Location
+    with type key := key
+     and type hash := hash
+     and type root_hash := hash
+     and type account := account
+
+  val cast : (module Base_intf with type t = 'a) -> 'a -> witness
```
Comment (imeckler) [inline src/lib/merkle_ledger/any_ledger.ml]: why cast?

File: src/lib/merkle_ledger_tests/test_stubs.ml [hunk 9]
```diff
@@ -141,7 +150,7 @@ module Key = struct
 
   let empty = Account.empty.public_key
 
-  let to_string = Format.sprintf !"%{sexp: T.t}"
+  (*   let to_string = Format.sprintf !"%{sexp: T.t}" *)
```
Comment (imeckler) [inline src/lib/merkle_ledger_tests/test_stubs.ml]: delete

File: src/lib/merkle_mask/masking_merkle_tree.ml [hunk 10]
```diff
@@ -327,18 +386,24 @@ struct
...
+         we sort mask locations in reverse order, potentially allowing reuse of locations
+      *)
+      let rev_sorted_mask_locations =
+        List.sort mask_locations ~compare:(fun loc1 loc2 ->
+            Location.compare loc2 loc1 )
+      in
+      List.iter rev_sorted_mask_locations
+        ~f:(remove_account_and_update_hashes t)
 
+    (* TODO : should destroy do a "commit" ? *)
```
Comment (imeckler) [inline src/lib/merkle_mask/masking_merkle_tree.ml]: should this be answered before merging?

File: src/lib/merkle_mask/masking_merkle_tree.ml [hunk 11]
```diff
@@ -392,28 +450,41 @@ struct
...
+        match Base.location_of_key (get_parent t) key with
+        | Some location -> Ok (`Existed, location)
+        | None ->
+            (* not in parent, create new location *)
+            let maybe_location =
+              match t.current_location with
+              | None -> Some first_location
+              | Some loc -> Location.next loc
+            in
+            if not (Option.is_some maybe_location) then
```
Comment (imeckler) [inline src/lib/merkle_mask/masking_merkle_tree.ml]: `match`

File: src/lib/snark_keys/gen_keys/jbuild [hunk 12]
```diff
@@ -14,8 +14,9 @@
       core
       ledger_builder
     ))
+   (link_flags (-cclib -lrocksdb))
    (preprocessor_deps ("../../../config.mlh"))
    (preprocess (pps (ppx_jane ppxlib.metaquot ppxlib.runner)))
-   (flags (-w -40 -g -warn-error +3+4+6+8+9+10+11+12+14+16+20+21+23+24+26+27+29+32..39+56-58+59+60+61))
+   (flags (-w -40 -g -warn-error +3+4+6+8+9+10+11+12+14+16+20+21+23+24+26+27+29+32..39+56-58+59+60+61 -cclib -lrocksdb))
```
Comment (imeckler) [inline src/lib/snark_keys/gen_keys/jbuild]: we can remove this yeah?

File: .gitignore (hunk 1 above)
Comment (bkase) [reply .gitignore]: this is the clone from an install script, we do not modify it

File: src/lib/coda_base/ledger.ml (hunk 5 above)
Comment (bkase) [reply src/lib/coda_base/ledger.ml]: We just lifted these functions out of a functor

File: src/lib/ledger_builder_controller/ledger_builder_controller.ml (hunk 6 above)
Comment (bkase) [reply src/lib/ledger_builder_controller/ledger_builder_controller.ml]: This is dead code now that we're not merging to master

File: src/lib/merkle_ledger/any_ledger.ml (hunk 7 above)
Comment (bkase) [reply src/lib/merkle_ledger/any_ledger.ml]: It's a type that witnesses the base ledger signature

File: src/lib/merkle_ledger/any_ledger.ml (hunk 8 above)
Comment (bkase) [reply src/lib/merkle_ledger/any_ledger.ml]: It's casting an 'a to a witness

File: src/lib/merkle_ledger/any_ledger.ml (hunk 7 above)
Comment (imeckler) [reply src/lib/merkle_ledger/any_ledger.ml]: ugh ok

**Comment by bkase** [comment]
Replaced by #1254

**Pull request closed without merging**